cluster/images/etcd/migrate/integration_test.go

```diff
@@ -0,0 +1,237 @@
+	peers := []string{}
+	for i := 0; i < memberCount; i++ {
+		memberName := fmt.Sprintf("%s-%d", name, i)
+		peerPort := uint64(2380 + i*10000)
+		peer := fmt.Sprintf("%s=%s://127.0.0.1:%d", memberName, protocol, peerPort)
+		peers = append(peers, peer)
+	}
+	initialCluster := strings.Join(peers, ",")
+
+	extraArgs := ""
+	if protocol == "https" {
+		extraArgs = getOrCreateTLSPeerCertArgs(t)
+	}
+
+	cfgs := []*EtcdMigrateCfg{}
+	for i := 0; i < memberCount; i++ {
+		memberName := fmt.Sprintf("%s-%d", name, i)
+		peerURL := fmt.Sprintf("%s://127.0.0.1:%d", protocol, uint64(2380+i*10000))
+		cfg := &EtcdMigrateCfg{
+			binPath:           "/usr/local/bin",
+			name:              memberName,
+			initialCluster:    initialCluster,
+			port:              uint64(2379 + i*10000),
+			peerListenUrls:    peerURL,
+			peerAdvertiseUrls: peerURL,
+			etcdDataPrefix:    "/registry",
+			ttlKeysDirectory:  "/registry/events",
+			supportedVersions: testSupportedVersions,
+			dataDirectory:     fmt.Sprintf("/tmp/etcd-data-dir-%s", memberName),
+			etcdServerArgs:    extraArgs,
+		}
+		cfgs = append(cfgs, cfg)
+	}
+	return cfgs
+}
+
+func getOrCreateTLSPeerCertArgs(t *testing.T) string {
+	spec := TestCertSpec{
+		host: "localhost",
+		ips:  []string{"127.0.0.1"},
+	}
+	certDir := "/tmp/certs"
+	certFile := filepath.Join(certDir, "test.crt")
+	keyFile := filepath.Join(certDir, "test.key")
+	err := getOrCreateTestCertFiles(certFile, keyFile, spec)
+	if err != nil {
+		t.Fatalf("failed to create server cert: %v", err)
+	}
+	return fmt.Sprintf("--peer-client-cert-auth --peer-trusted-ca-file=%s --peer-cert-file=%s --peer-key-file=%s", certFile, certFile, keyFile)
+}
+
+type TestCertSpec struct {
+	host       string
+	names, ips []string // in certificate
+}
+
+func getOrCreateTestCertFiles(certFileName, keyFileName string, spec TestCertSpec) (err error) {
+	if _, err := os.Stat(certFileName); err == nil {
+		if _, err := os.Stat(keyFileName); err == nil {
+			return nil
+		}
+	}
+
+	certPem, keyPem, err := generateSelfSignedCertKey(spec.host, parseIPList(spec.ips), spec.names)
+	if err != nil {
+		return err
+	}
+
+	os.MkdirAll(filepath.Dir(certFileName), os.FileMode(0755))
+	err = ioutil.WriteFile(certFileName, certPem, os.FileMode(0755))
+	if err != nil {
+		return err
+	}
+
+	os.MkdirAll(filepath.Dir(keyFileName), os.FileMode(0755))
+	err = ioutil.WriteFile(keyFileName, keyPem, os.FileMode(0755))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func parseIPList(ips []string) []net.IP {
+	var netIPs []net.IP
+	for _, ip := range ips {
+		netIPs = append(netIPs, net.ParseIP(ip))
+	}
+	return netIPs
+}
+
+// generateSelfSignedCertKey creates a self-signed certificate and key for the given host.
+// Host may be an IP or a DNS name
+// You may also specify additional subject alt names (either ip or dns names) for the certificate
+func generateSelfSignedCertKey(host string, alternateIPs []net.IP, alternateDNS []string) ([]byte, []byte, error) {
```

Comment:
I need to decide what do to here. This is a near copy of the code under `staging/src/k8s.io/apiserver/pkg/server/options/service_testing.go`. The only difference is that `x509.ExtKeyUsageClientAuth` is set.  I wanted to avoid checking in certs and keys to git (both because they can expire and then need to be regenerated, and also because their presence in source control can raise security concerns) so instead we're generating test certs and keys here.